### PR TITLE
New scoring metrics for "cross_validation_across_time"

### PR DESCRIPTION
**Pull request by sianna338**
The function **cross_validation_across_time** can now use different scoring algorithms to determine classifier performance. 
The parameter "metric" is added which accepts either a string keyword, refering to one of the scoring functions from sklearn.metrics, or alternatively a user-defined scoring function. Default is "accuracy".

### CODE REVIEW

File: decoding.py [hunk 1]
```diff
@@ -154,7 +154,8 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
                                  n_jobs=-2, plot_confmat=False, title_add="",
                                  ex_per_fold=2, simulate=False, subj="",
                                  tmin=-0.1, tmax=0.5, sfreq=100,
-                                 return_probas=False,
+                                 return_probas=False, metric='accuracy',
+                                 metric_kwargs={}, proba=True,
```
Comment (skjerns) [inline decoding.py]: We now have `proba` and `return_proba` - the seem to me as to do the same thing, no?

File: decoding.py [hunk 2]
```diff
@@ -247,7 +263,7 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
                 test_x=test_x[:, :, start],
                 neg_x=neg_x,
                 clf=clf,
-                proba=True
+                proba=proba
```
Comment (skjerns) [inline decoding.py]: shouldn't `proba` always be True? In what case should it be `False` ?

File: decoding.py [hunk 3]
```diff
@@ -258,8 +274,32 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
         all_probas[idxs_test] = probas
 
         preds = np.argmax(probas, -1)
-
-        accuracy = (preds == test_y[:, None]).mean(axis=0)
+        
+        if metric == "accuracy": 
+            accuracy = (preds == test_y[:, None]).mean(axis=0)
+        else:
```
Comment (skjerns) [inline decoding.py:257]: you can also do this using the `top_k_accuracy_score(k=1)` :) makes it a bit cleaner

File: decoding.py [hunk 4]
```diff
@@ -258,8 +274,32 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
...
+            if metric_kwargs:
+                kwargs = {}
+                for k, v in metric_kwargs.items():
+                    if k in sig.parameters:
+                        kwargs[k] = v
+                        
+            # need to loop over timepoints 
+            accuracy = np.zeros(time_max)
+            for t in list(range(0, time_max)): 
+                accuracy[t] = func(data_y, all_probas[:,t], **kwargs)
```
Comment (skjerns) [inline decoding.py]: maybe rename variable from accuracy to `score`, as it now can be any kind of result?

File: decoding.py [hunk 5]
```diff
@@ -258,8 +274,32 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
...
+            else:
+                raise TypeError("metric must be 'accuracy', a sklearn.metrics name (str), or a callable.")
+    
+            sig = inspect.signature(func)
+            # add any extra parameters that are not preds and data_y
+            if metric_kwargs:
+                kwargs = {}
+                for k, v in metric_kwargs.items():
+                    if k in sig.parameters:
+                        kwargs[k] = v
```
Comment (skjerns) [inline decoding.py]: I would throw an error if the parameter is not part of the called function! otherwise we silently drop it and the user might not realize his kwarg had no effect.
```suggestion
            if metric_kwargs:
                sig = inspect.signature(func)
                missing_kwargs = set(metric_kwargs).difference(sig.parameters)
                if missing_kwargs:
                    raise ValueError(f'The following metric_kwargs were given but are not part of the function signature {missing_kwargs} ')
```

and then later simply pass metric_kwargs` instead of `kwargs` to the function call

File: decoding.py (hunk 4 above)
Comment (skjerns) [inline decoding.py]:
```suggestion
                accuracy[t] = func(data_y, all_probas[:,t], **metric_kwargs)
```
after resolving the above comments

File: decoding.py [hunk 6]
```diff
@@ -258,8 +274,32 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
...
+                        
+            # need to loop over timepoints 
+            accuracy = np.zeros(time_max)
+            for t in list(range(0, time_max)): 
+                accuracy[t] = func(data_y, all_probas[:,t], **kwargs)
 
         # Create a temporary DataFrame for the current fold
         df_temp = pd.DataFrame(
```
Comment (skjerns) [inline decoding.py:319]: here you should probably also change the metric name in the dataframe.

Additionally, adding the kwargs to the dataframe might be useful, but only add it if there were any given e.g.

`"metric_kwargs": str(metric_kwargs)`

File: decoding.py (hunk 1 above)
Comment (sianna338) [reply decoding.py]: you're right it doesnt make sense

Comment (sianna338) [inline decoding.py]:
```suggestion
                                 metric_kwargs={}, 
```

File: decoding.py (hunk 2 above)
Comment (sianna338) [reply decoding.py]:
```suggestion
                proba=return_proba
```

Comment (sianna338) [reply decoding.py]: some sklearn.metrics functions (like for example: "balanced_accuracy_score") expect predicted labels as input instead of the classifier probability per class. But there might be a more elegant way to solve this, since now the user has to check this requirement before selecting the scoring function

File: decoding.py (hunk 4 above)
Comment (sianna338) [reply decoding.py]:
```suggestion
            score= np.zeros(time_max)
            for t in list(range(0, time_max)): 
                score[t] = func(data_y, all_probas[:,t], **kwargs)
```

File: decoding.py (hunk 2 above)
Comment (skjerns) [reply decoding.py]: Indeed!

Have a look at `sklearn.metrics.get_scorer` and `sklearn.metrics.get_scorer` this can be used in this case quite elegantly to get a scoring method that works on all kinds of inputs and automatically does that afaik :)

File: decoding.py (hunk 4 above)
Comment (skjerns) [reply decoding.py]: Don't just suggest things, implement them locally and push to the branch :D

Comment (sianna338) [reply decoding.py]: haha sorry, just trying out what happens when I click the buttons

**Comment by skjerns** [comment]
@sianna338 let me know when I should review again, and please comment and/or mark the conversations above as "resolved", so I know that it has been taken care of

File: decoding.py (hunk 2 above)
Comment (sianna338) [reply decoding.py]: I checked out sklearn.metrics.get_scorer and make_scorer, but they don’t seem to fit this case.
get_scorer only returns predefined scorers and doesn’t return the expected input type (no response_method).
make_scorer lets you specify response_method (e.g. .predict_proba) but seems to work only with tools like cross_val_score or GridSearchCV.

File: .gitignore [hunk 7]
```diff
@@ -160,3 +160,5 @@ cython_debug/
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 #.idea/
+decoding_test-1.py
+prepare_dat.py
```
[the remark this answers is not in the document]
Comment (skjerns) [reply .gitignore:164]: @sianna338 

generally dont add them to the gitignore, but rather just dont add them to the commit instead! but shuold be no problem

File: decoding.py [hunk 8]
```diff
@@ -18,12 +18,14 @@
 import pandas as pd
 import json
 import mne
+import misc
+import numpy as np
```
[the remark this answers is not in the document]
Comment (skjerns) [reply decoding.py:22]: @sianna338 please fix

File: decoding.py [hunk 9]
```diff
@@ -234,39 +245,84 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
...
+        
+            second_param = inputs_names[1]
+            prob_indicators = ['y_score', 'probas_pred', 'y_proba']  
+            pred_indicators = ['y_pred', 'labels_pred']
+            if any(indicator in second_param for indicator in prob_indicators):
+                needs_probas=True
+            elif any(indicator in second_param for indicator in pred_indicators):
+                needs_probas=False
+            else:
+                print("determining response method not possible")
```
Comment (skjerns) [inline decoding.py:279]: you can get the needed response method by using the scorers that are returned from `get_scorer`

e.g.

```python
if isinstance(metric, str):
    # assume that it is a sklearn.metric scorer
    scorer = sklearn.metrics.get_scorer()
    # dont use scorer directly, but extract methods from it
    response_method = scorer._response_method
    metric_func= scorer._score_func
else:
    # simply assume the custom function will take predict_proba
    # if not, the user can simply write a wrapper around his function
    # that converts proba to class labels.
    response_method = 'predict_proba'
    metric_func= metric
```

then later you can write something like

```python
if response_method=='predict':
    ... code to use with labels
    metric_func(preds, y_true)
elif response_method=='predict_proba':
    ... code that uses probabilities to calculate
    metric_func(probabilities, y_true)
```

File: decoding.py [hunk 10]
```diff
@@ -234,39 +245,84 @@ def cross_validation_across_time(data_x, data_y, clf, add_null_data=False,
...
+        
+        results_swp = np.swapaxes(results_preds, 0, 1)
+        
+        # store results for each fold together
+        all_results[idxs_test] = results_swp
+        
+        # convert probabilities to prediction if necessary 
+        if not needs_probas:
+           preds_idx = np.argmax(all_results, axis=2)              
+           preds_lbl = labels[preds_idx]
```
Comment (skjerns) [inline decoding.py:307]: move this down inside the other `else` clause, so it is where it's actually needed
